### PR TITLE
Fix null unboxing of primitives (fixes #582)

### LINKED ISSUE #582: null unboxing is broken because the backend does not intercept the unbox calls
This code:
```scala
object Test {
  def main(args: Array[String]): Unit = {
    val d: Double = null.asInstanceOf[Double]
    println(d)
  }
}
```
Is transformed both in `scalac` and dotty into:
```scala
def main(args: String[]): Unit = {
  val d: Double = scala.Double.unbox(null)
  println(scala.Double.box(d))
}
```
But this is just an illusion: in `scalac`, the `unbox` call is actually replaced by `BoxesRunTime.unboxToDouble(null)`. Dotty does not do this replacement and so the call fails with a `NullPointerException`.

(I still wonder why there even is an implementation in Double.scala since it's both unused and incorrect: https://github.com/scala/scala/blob/2.11.x/src/library/scala/Double.scala#L244)

### PR DESCRIPTION
**Pull request by smarter**
Erasure#isUnbox was incorrect: the unbox methods are defined on the
value classes, not the boxed classes.

Review by @DarkDimius.

### CODE REVIEW

File: tests/pos/nullAsInstanceOf.scala [hunk 1]
```diff
@@ -1,11 +0,0 @@
-object Test {
```
Comment (DarkDimius) [inline tests/pos/nullAsInstanceOf.scala]: for future reference, you can just write `extends dotty.runtime.LegacyApp` and body will be executed.

Comment (smarter) [inline tests/pos/nullAsInstanceOf.scala]: Do we want to do this for new tests to? I thought `LegacyApp` was only a compatiblity hack.

Comment (DarkDimius) [inline tests/pos/nullAsInstanceOf.scala]: That's a compatibility adaptor that works only if you do not inspect arguments. And that is the current situation with run tests.

**Comment by DarkDimius** [comment]
LGTM.

**Comment by smarter** [comment]
Note that `null.asInstanceOf[Unit]` is erased to `null` instead of `scala.Unit.unbox(null)`, so it prints `null` instead of `()`. `scalac` has the same behavior, I don't know if we want to fix that or not.

**Pull request merged**